File: go/problems/trie_prefix_tree.go
package problems

// https://leetcode.com/problems/implement-trie-prefix-tree
type Trie struct {
  isEnd bool
  children []*Trie
}

/** Initialize your data structure here. */
func NewTrie() *Trie {
  return &Trie{
      isEnd: false,
      children: make([]*Trie, 26),
  }
}

/** Inserts a word into the trie. */
func (t *Trie) Insert(word string)  {
  curr := t
  for i := 0; i < len(word); i++ {
    c := int(word[i] - 'a')
    if curr.children[c] == nil {
      curr.children[c] = NewTrie()
    }
    curr = curr.children[c]
  }
  curr.isEnd = true
}

/** Returns if the word is in the trie. */
func (t *Trie) Search(word string) bool {
  cur := t
  for i := 0; i < len(word); i++ {
    c := int(word[i] - 'a')
    if cur.children[c] == nil {
      return false
    }
    cur = cur.children[c]
  }

  return cur.isEnd
}

/** Returns if there is any word in the trie that starts with the given prefix. */
func (t *Trie) StartsWith(prefix string) bool {
  cur := t
  for i := 0; i < len(prefix); i++ {
    c := int(prefix[i] - 'a')
    if cur.children[c] == nil {
      return false
    }
    cur = cur.children[c]
  }

  return true
}

File: rust/src/problems/trie_prefix_tree.rs
#[derive(Default)]
struct Trie {
  is_end: bool,
  children: [Option<Box<Trie>>; 26],
}

pub trait TrieOps {
  fn new() -> Self;
  fn insert(&mut self, word: String);
  fn search(&self, word: String) -> bool;
  fn starts_with(&self, prefix: String) -> bool;
}

impl TrieOps for Trie {
  fn new() -> Self {
    return Default::default();
  }

  fn insert(&mut self, word: String) {
    let mut curr = self;
    for c in word.chars() {
      let val = &curr.children[c as usize - 'a' as usize];
      match val {
        None => curr.children[c as usize - 'a' as usize] = Some(Box::new(Trie::new())),
        Some(_) => ()
      }
      curr = curr.children[c as usize - 'a' as usize].as_mut().unwrap();
    }
    curr.is_end = true;
  }

  fn search(&self, word: String) -> bool {
    let mut cur = self;
    for c in word.chars()
    {
      match cur.children[c as usize - 'a' as usize].as_ref()
      {
        None => return false,
        Some(n) => cur = n.as_ref(),
      }
    }
    return cur.is_end;
  }

  fn starts_with(&self, prefix: String) -> bool {
    let mut cur = self;
    for c in prefix.chars()
    {
      match cur.children[c as usize - 'a' as usize].as_ref()
      {
        None => return false,
        Some(n) => cur = n.as_ref(),
      }
    }
    return true;
  }
}